bindings/java/src/org/sleuthkit/datamodel/OsAccountManager.java

```diff
@@ -55,121 +54,125 @@
 
+		if (realmScope == null) {
+			throw new IllegalArgumentException("RealmScope cannot be null. Use UNKNOWN if scope is not known.");
+		}
+		if (referringHost == null) {
+			throw new IllegalArgumentException("A referring host is required to create an account.");
+		}
+		
 		// ensure at least one of the two is supplied - unique id or a login name
-		if (Strings.isNullOrEmpty(uniqueAccountId) && Strings.isNullOrEmpty(loginName)) {
+		if (Strings.isNullOrEmpty(sid) && Strings.isNullOrEmpty(loginName)) {
 			throw new IllegalArgumentException("Cannot create OS account with both uniqueId and loginName as null.");
 		}
-		
-		if (Strings.isNullOrEmpty(realmName)) {
-			throw new IllegalArgumentException("Realm name is required to create an OS account.");
+		// Realm name is required if the sid is null. 
+		if (Strings.isNullOrEmpty(sid) && Strings.isNullOrEmpty(realmName)) {
+			throw new IllegalArgumentException("Realm name or SID is required to create a Windows account.");
 		}
 
-		if (!Strings.isNullOrEmpty(uniqueAccountId)) {
-			Optional<OsAccount> osAccount = this.getOsAccountByUniqueId(uniqueAccountId, host, transaction.getConnection());
-			if (osAccount.isPresent()) {
-				return osAccount;
-			}
-		}
-		
-		if (!Strings.isNullOrEmpty(loginName)) {
-			// first get the realm 
-			Optional<OsAccountRealm> realm = db.getOsAccountRealmManager().getRealmByName(realmName, host, transaction);
+		Optional<OsAccountRealm> realm = Optional.empty();
+		try (CaseDbConnection connection = this.db.getConnection();) {
+
+			// get the realm with given name
+			realm = db.getOsAccountRealmManager().getWindowsRealm(sid, realmName, referringHost, connection);
 			if (!realm.isPresent()) {
-				throw new TskCoreException(String.format("No realm found with name %s", realmName));
+				// realm was not found, create it.
+				realm = Optional.of(db.getOsAccountRealmManager().createWindowsRealm(sid, realmName, referringHost, realmScope));
 			}
-
-			return getOsAccountByLoginName(loginName, realm.get(), transaction.getConnection());
-		}
 		
-		return Optional.empty();
+			// try to create account
+			try {
+				return createOsAccount(sid, loginName, realm.get(), OsAccount.OsAccountStatus.UNKNOWN, connection);
+			} catch (SQLException ex) {
+
+				// Create may fail if an OsAccount already exists. 
+				Optional<OsAccount> osAccount;
+
+				// First search for account by uniqueId
+				if (!Strings.isNullOrEmpty(sid)) {
+					osAccount = getOsAccountByUniqueId(sid, realm.get());
+					if (osAccount.isPresent()) {
+						return osAccount.get();
+					}
+				}
+
+				// search by loginName
+				if (!Strings.isNullOrEmpty(loginName)) {
+					osAccount = getOsAccountByLoginName(loginName, realm.get());
+					if (osAccount.isPresent()) {
+						return osAccount.get();
+					}
+				}
+
+				// create failed for some other reason, throw an exception
+				throw new TskCoreException(String.format("Error creating OsAccount with loginName = %s", loginName), ex);
```

Comment:
Are we throwing away some additional information here that might help debugging or is login name sufficent?